rl4co/envs/routing/cvrp/env.py

```diff
@@ -0,0 +1,215 @@
+from typing import Optional
+
+import torch
+
+from tensordict.tensordict import TensorDict
+from torchrl.data import (
+    BoundedTensorSpec,
+    CompositeSpec,
+    UnboundedContinuousTensorSpec,
+    UnboundedDiscreteTensorSpec,
+)
+
+from rl4co.data.utils import load_npz_to_tensordict
+from rl4co.envs.common.base import RL4COEnvBase
+from rl4co.utils.ops import gather_by_index, get_tour_length
+from rl4co.utils.pylogger import get_pylogger
+
+from .generator import CVRPGenerator
+from .render import render
+
+log = get_pylogger(__name__)
+
+
+class CVRPEnv(RL4COEnvBase):
+    """Capacitated Vehicle Routing Problem (CVRP) environment.
+    
+    Observations:
+        - location of the depot.
+        - locations and demand of each customer (city).
+        - current location of the vehicle.
+        - the remaining capacity of the vehicle,
+
+    Constraints:
+        - the tour starts and ends at the depot.
+        - each city must be visited exactly once.
+        - the vehicle cannot visit cities exceed the remaining capacity.
+        - the vehicle can return to the depot to refill the capacity.
+
+    Finish Condition:
+        - the vehicle has visited all cities and returned to the depot.
+
+    Reward:
+        - (minus) the negative length of the path.
+
+    Args:
+        generator: CVRPGenerator instance as the data generator
+        generator_params: parameters for the generator
+    """
+
+    name = "cvrp"
+
+    def __init__(
+        self,
+        generator: CVRPGenerator = None,
+        generator_params: dict = {},
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        if generator is None:
+            generator = CVRPGenerator(**generator_params)
+        self.generator = generator
+        self._make_spec(self.generator)
+
+    def _step(self, td: TensorDict) -> TensorDict:
+        current_node = td["action"][:, None]
+
+        # Get demand for selected node
+        selected_demand = gather_by_index(td["demand"], current_node, squeeze=False)
+
+        # If not depot, add demand to used capacity; if depot, reset used capacity to 0
+        used_capacity = (td["used_capacity"] + selected_demand) * (current_node != 0).float()
+
+        # Update visited nodes
+        visited = td["visited"].scatter(-1, current_node, True)
+
+        done = visited.sum(-1, keepdim=True) == visited.size(-1)
+        reward = torch.zeros_like(done)
```

Comment:
This could be done, but I'm not sure how "reusable" this would be. In some routing envs, we don't need to visit all nodes (e.g. OP) or we can visit more than once (SDVRP). So modularizing this part might be challenging and perhaps lead to some hard-to-read code in terms of logic. But I'm open to be proven wrong here